components/capi/src/custom_writeable.rs

```diff
@@ -0,0 +1,94 @@
+use std::ffi::c_void;
+use std::{fmt, ptr};
+
+#[repr(C)]
+/// An object that can one can write UTF-8 strings to
+///
+/// This allows the C API to write to arbitrary kinds of objects, for example a
+/// C++ std::string or a char buffer.
+///
+/// [`icu4x_simple_writeable()`] can be used to write to a fixed-size char buffer.
+///
+/// # Safety invariants:
+///  - `flush()` and `grow()` will be passed `data` and the value should be valid for that
+///     `data` may be  null, however `flush()` and `grow()` must then be ready to receive it
+///  - `buf` must be `cap` bytes long
+///  - `grow()` must either return null or a valid buffer of at least the requested buffer size
+///  - Rust code must call `ICU4XCustomWriteable::flush()` before releasing to C
+pub struct ICU4XCustomWriteable {
+    /// Pointer to the actual object. While we're writing, we will write
+    /// directly to `buf` without updating `data`'s state.
+    data: *mut c_void,
```

Comment:
Suggestion: rename `data` to `context` and update the comment to reflect that this is not necessarily a pointer to anything.  It could be `nullptr`, or it could be a pointer to a structured object like `std::string`.  Its only purpose is to be passed as an argument to the `flush()` and `grow()` functions.